### DISCUSSION TITLE
Production Build Not Displaying in Browser

### DISCUSSION THREAD

**Discussion by peniswafflexd** [payloadcms/payload#554] [Q&A]
So I've basically got payload set up the way I would like now. However, after trying to make a production build:
`yarn build` || `npm run build`
and then serving said build:
`serve -s build -l 3000`
Nothing will display in the browser, it seems as though there is nothing there at all, none of the css or js files are showing when I try navigate to the, although I'm not getting a 404 or other error status code. There is just a blank page.

I am serving the build package to the same port and address, so it shouldn't be an issue with any of the hardcoded environment variables.

I can't figure out for the life of me what the cause is, here is my build config and server.js files:
** config **
```js
export default buildConfig({
    serverURL: 'http://localhost:3000',
    admin: {
        user: Users.slug,
        scss: resolve(__dirname, 'vars.scss'),
        css: resolve(__dirname, 'stylesheet.scss'),
    },
    collections: [
        Categories,
        Posts,
        Tags,
        Users,
        HeaderLinks,
        Pages,
        SocialLinks,
        Media,
        AboutCard,
        EmailList
    ],
    upload: {
        limits: {
            fileSize: 500000000, // 5MB, written in bytes
        }
    },
    csrf: [
        'http://localhost:3001'
    ],
    cors: [
        'http://localhost:3001'
    ],
});
```

** server.js  **

```js
const transport = nodemailer.createTransport({
  host: process.env.SMTP_HOST,
  port: 25,
  tls:{
    rejectUnauthorized: false
  }
});

transport.use('compile', inlineBase64())
// Initialize Payload
payload.init({
  secret: process.env.PAYLOAD_SECRET,
  mongoURL: process.env.MONGODB_URI,
  express: app,
  onInit: () => {
    payload.logger.info(`Payload Admin URL: ${payload.getAdminURL()}`);
  },
  email: {
    fromName: "Devon",
    fromAddress: "Devon <devon@myEmailServer>",
    transport
  }
});

app.get('/unsubscribe/:emailID', (req, res) => {
  fetch(process.env.HOST+`/api/emaillist/${req.params.emailID}`, {
    method: 'DELETE'
  }).then(() => {
    res.status(200)
    res.send("You've successfully unsubscribed")
  }).catch(err => {
    res.status(503)
    res.send({error: err})
  })
})

app.listen(process.env.PORT || 3000)
```

**Comment by jmikrut** [comment]
Hey @peniswafflexd — 

Few questions for you:

1. What route are you hitting in your browser to receive a blank page?
1. What does your build script do? Can you paste your `package.json`, or at least the `scripts` from it?

There is nothing wrong with your Payload config, or your server, from what I can see. It's probably related to your `serve` command or your `build` command.

Let me know!

**Reply by peniswafflexd** [reply]
Hey sure,
Any route actually... /api or /admin 
```js
"scripts": {
    "dev": "nodemon",
    "build": "payload build",
    "serve": "NODE_ENV=production node server.js"
  },
```

And Ive just been running it with the [serve](https://www.npmjs.com/package/serve) package.

**Reply by DanRibbens** [reply]
I think serve needs to be "NODE_ENV=production node dist/server.js" since the build step will put your app in /dist.

**Reply by peniswafflexd** [reply]
Just to be clear this is what happens when trying to browse to the admin panel on the production build
![image](https://user-images.githubusercontent.com/28529309/166688678-cb84b46a-34ac-4dee-9a30-6ad487fd8828.png)

**Reply by peniswafflexd** [reply]
> I think serve needs to be "NODE_ENV=production node dist/server.js" since the build step will put your app in /dist.

I gave that a try aswell and no change, im racking my brain trying to figure out what could cause it. It could be the serve package that I am using. Is there anything else you can recommend to try hosting the build locally

**Reply by DanRibbens** [reply]
I'm not familiar with [serve](https://www.npmjs.com/package/serve). From the readme under **API**
> The core of serve is [serve-handler](https://github.com/vercel/serve-handler), which can be used as middleware in existing HTTP servers: ...

Can you try changing your server.js like their example?

**Reply by peniswafflexd** [reply]
Hmm I've only ever used to it serve static files, so I wouldnt really know how to go about that, and not sure it would be needed. 
I just tried with python and it gave the same result sadly. 

I'm only trying to serve it so I can make sure it works before putting on the production server, which will be using apache virtual hosts to serve it

**Reply by DanRibbens** [reply]
What do you see in the network tab?
There has to be a cors error or something we're missing.

**Reply by peniswafflexd** [reply]
Network tab is clean, apparantly its getting 200 accepted accross the board 
![image](https://user-images.githubusercontent.com/28529309/166698216-0eff2e66-5323-4077-a321-1418b2956e9b.png)

**Reply by peniswafflexd** [reply]
Whoops thats my bad, its not getting 200's its getting 404 errors. Had single page mode running on serve
![image](https://user-images.githubusercontent.com/28529309/166699088-aa066836-777c-4c75-8822-de07c8bfc27a.png)

**Reply by peniswafflexd** [reply]
I'm sure you're a busy man @DanRibbens but if you want I could make my repo public for a bit and you could try build and serve it and see if you have the same errors. 
It might be due to something in my environment. I'm running that latest release of Nodejs and npm. I'm running on the m1 arm architecture (not that I think that would matter much). Theres only 7 dependencies in my package.json file.

I'm just not sure if this is a me issue or if it is a payload issue at this point

**Reply by DanRibbens** [reply]
Does your other route work that you've added? 
> DELETE /unsubscribe/:emailID

What does your console output show when you serve?

**Reply by DanRibbens** [reply]
> I'm sure you're a busy man @DanRibbens but if you want I could make my repo public for a bit and you could try build and serve it and see if you have the same errors.

Sure, if you want to share your repo with me or make it public I can troubleshoot it. That is probably faster than all the back and forth.

**Reply by peniswafflexd** [reply]
I just added you as a collaborator to the repo, you should've got an invite

**Reply by DanRibbens** [reply]
![image](https://user-images.githubusercontent.com/6434612/166705348-855fd901-df1b-4102-bdf8-3281ecd3225e.png)

From a windows machine it is working using `yarn serve`. I don't have an m1 mac to try. I'm testing on wsl and npm next.

**Reply by peniswafflexd** [reply]
Ahh, so yes, I can serve it fine using yarn serve. My issue is that I created a production build, ‘yarn build’ so I can serve the static files directly, opposed to running a service. 
basically I was hoping not to have to actually run node on my server and just serve the static files via Apache virtual hosts

**Reply by DanRibbens** [reply]
That would be okay to serve the /admin route statically.

I can see that you were able to build that, it is in your repo here: https://github.com/peniswafflexd/payload_blog/blob/main/build/index.html

You'll still need node for the API of course.

**Reply by peniswafflexd** [reply]
Ahh I see… well it looks like I’ve had us chasing down a rabbit hole for no reason then haha sorry about that, I obviously didn’t think too hard when I thought that I’d be able to host the entire backend statically….

Well I appreciate all the help nonetheless!

**Reply by DanRibbens** [reply]
Ah, no worries!

**Comment by DanRibbens** [comment]
Hi @peniswafflexd,
Try removing the serverURL. I think the production it won't be on port 3000 and this is likely the issue. Payload can start up without the serverURL property so just take that out and see what happens.

**Reply by peniswafflexd** [reply]
Sure I can take it out, but it was being hosted on port 3000 still. Ill try it out

**Reply by peniswafflexd** [reply]
I'm actually getting build errors if i try to remove the serverURL:
```
devon@Devons-MacBook-Pro personal-blog % npm run build

> payload-blog@1.0.0 build
> payload build

[13:01:07] ERROR (payload): There were 2 errors validating your Payload config
[13:01:07] ERROR (payload): 1: "serverURL" is required
[13:01:07] ERROR (payload): 2: "csrf[1]" must not be a sparse array item
```

**Reply by DanRibbens** [reply]
In a more recent version of Payload serverURL is no longer required, after you upgrade it should build.

**Reply by peniswafflexd** [reply]
I just updated to v0.16.2 and removed the serverURL and it did build, but sadly it didnt seem to fix the issue
